ITE/10/README.adoc

```diff
@@ -0,0 +1,265 @@
+= ITE-10: Layouts for in-toto Attestations
+:source-highlighter: pygments
+:toc: preamble
+:toclevels: 2
+ifdef::env-github[]
+:tip-caption: :bulb:
+:note-caption: :information_source:
+:important-caption: :heavy_exclamation_mark:
+:caution-caption: :fire:
+:warning-caption: :warning:
+endif::[]
+
+.Metadata
+[cols="2"]
+|===
+| ITE
+| 10
+
+| Title
+| Layouts for in-toto Attestations
+
+| Sponsor
+| link:https://github.com/adityasaky[Aditya Sirish A Yelgundhalli]
+
+| Status
+| Draft
+
+| Type
+| Standards
+
+| Created
+| 2023-01-07
+
+|===
+
+[[abstract]]
+== Abstract
+
+A previous in-toto Enhancement, ITE-6, presented the in-toto Attestation
+Framework. This framework introduced the idea of context-specific attestations
+with their own definitions. This ITE builds on ITE-6 and proposes several
+modifications to in-toto layouts that allow for defining and verifying policies
+over ITE-6 attestations.
+
+[[specification]]
+== Specification
+
+Note: this document assumes the reader is familiar with ITE-6 and the in-toto
+Attestation Framework.
+
+in-toto layouts have two key mechanisms for specifying policies: steps and
+inspections. Steps are used to reason about and express constraints regarding
+operations performed as part of the supply chain. Constraints are currently
+expressed using in-toto artifact rules against the materials and products of a
+step. Each step definition also declares the set of functionaries authorized to
+perform the step and the required number of functionaries that need to sign off
+on the step.
+
+Inspections, on the other hand, are executed during the verification workflow to
+complement in-toto's step verifications with extra checks. Typically,
+inspections invoke external commands or scripts shipped with the layout. They
+may also use artifact rules. Currently, inspections are used for constraints
+in-toto layouts do not support. For example, in-toto links can record arbitrary
+information in their opaque environment and byproducts fields. To validate their
+contents, supply chain owners must rely on custom verification scripts invoked
+as inspections.
+
+=== Classifying Predicates
+
+Prior to ITE-6, all steps were validated against in-toto links. Inspection
+executions also resulted in the generation of in-toto links that were then
+validated. With the introduction of the in-toto Attestation Framework, it is
+necessary to rework the layout to support verifying a variety of _predicates_.
+As the framework is designed to generate claims pertaining to how software is
+produced, each predicate type can be mapped to a step in the software supply
+chain. However, it is first necessary to understand the nature of each step, and
+therefore its predicate type. in-toto was originally designed with the
```

Comment:
This seems to imply there is always a 1:1 correspondence between steps and predicate types. I'm not sure this is strictly true, as it's possible for a single step to generate multiple attribute type predicates along with a link type predicate. For example, a build step may simultaneously generate SLSA provenance, a runtime trace predicate about the build system, and a SCAI predicate about the build platform. All three of these predicates are tied to a particular build run, so generating them disjointly wouldn't necessarily give the expected results.

So, I might amend this description to be less restrictive about how many or which classes of predicates can be generated by a particular step.

Reply:
Right, because I don't think it's actually true: I remember Aditya pointing out that a step could use an attestation that fits one of many predicates (but here, the technical challenge is figuring out whether the [data types](https://en.wikipedia.org/wiki/Unification_(computer_science)#Application:_type_inference) are compatible, which is a fairly standard problem in computer science, but will add a lot of complexity to any in-toto verifier).

Reply:
I'm not sure I follow... what about my comment evoked a relationship to type inference?

Reply:
Oh, right, sorry, this is not strictly the predicate _data type_, but rather whether it's the modify-materials-products or the opposite (e.g., source code review) model. We really should use a very different term than "predicate type" for this. Nevertheless, my comment remains for multiple predicate data types per step, which I've advised Aditya about.

Reply:
I do have a draft of steps supporting multiple predicates with policies for each. I'll sketch it out some more and update the document.

Reply:
I've updated it with support for multiple predicates per step.

---

As such, this ITE proposes the following schema for step declarations.

```yaml
name: string
command: string
expectedPredicates: list of ExpectedStepPredicates
```

Each `ExpectedStepPredicate` object has the following schema.

```yaml
predicateTypes: list of TypeURIs
expectedMaterials: list of constraints
expectedProducts: list of constraints
expectedAttributes: list of constraints
functionaries: list of strings
threshold: int
```